### PR TITLE
Connection impersonation FE

### PR DESCRIPTION
**Pull request by alxnddr**
Epic: https://github.com/metabase/metabase/issues/29491

### Description

Updates and connects the impersonation UI to the backend.

### How to verify

There is a known bug. If you configure an impersonation for the first time, it will reset your native query editing permission from "Yes" to "No". I left a couple of [FIXME](https://github.com/metabase/metabase/pull/30867/files#diff-255dc4e7aa757a489ef9b7b41fe27e5ba89f0e1a14ab22659a4496327c59b55cR146) in the specs code.

- Run Metabase EE
- Connect postgres & snowflake databases
- Create roles in postgres and snowflake that limit access to certain tables
- Go to Admin -> People, create a new user, assign two attributes to the user: pg_role=*your_pg_role*, snowflake_role=*your_snowflake_role*
- Go to Admin -> Permissions, for "All users" set the data access to `Impersonated`, select corresponding attributes for pg and snowflake
- Login as the user
- Ensure users cannot see any data beyond what the roles give access to both for structured and native queries

### Demo

Screen Shot 2023-05-19 at 5 40 33 PM

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30867)

### CODE REVIEW

File: e2e/test/scenarios/permissions/impersonated.cy.spec.js
Comment (iethree) [inline e2e/test/scenarios/permissions/impersonated.cy.spec.js:1]: what do you think about adding a spec to this that tests actions permissions: i.e. a user role that has read permissions but not write?

File: frontend/src/metabase-types/store/state.ts [hunk 1]
```diff
@@ -26,7 +26,7 @@ export interface State {
   upload: FileUploadState;
 }
 
-export type Dispatch<T = any> = (action: T) => void;
+export type Dispatch<T = any> = (action: T) => any;
```
Comment (iethree) [inline frontend/src/metabase-types/store/state.ts:29]: is there any way to avoid `any` here?

**Review by iethree — APPROVED** [review]
🥇 

Looks great, tested with postgres. will test snowflake too, and follow up if I see anything different there. There's a couple lint issues to fix and one testing suggestion.

Also, I noticed some leakage that violates the impersonation, not sure if that's on the radar already. https://www.loom.com/share/d30207fb80ee4f75aa33231cf316e2d1

File: e2e/test/scenarios/permissions/impersonated.cy.spec.js
Comment (alxnddr) [reply e2e/test/scenarios/permissions/impersonated.cy.spec.js:1]: Good idea, will add in a separate PR

**Pull request merged**